monai/metrics/cumulative_average.py

```diff
@@ -9,64 +9,160 @@
+class CumulativeAverage:
     """
-    Cumulatively record data value and aggregate for the average value.
-    It supports single class or multi-class data, for example,
-    value can be 0.44 (a loss value) or [0.3, 0.4] (metrics of two classes).
-    It also supports distributed data parallel, sync data when aggregating.
-    For example, recording loss values and compute the overall average value in every 5 iterations:
+    A utility class to keep track of average values. For example during training/validation loop,
+    we need to accumulate the per-batch metrics and calculate the final average value for the whole dataset.
+    When training in multi-gpu environment, with DistributedDataParallel, it will average across the processes.
+
+    Example:
 
     .. code-block:: python
 
-        average = CumulativeAverage()
-        for i, d in enumerate(dataloader):
-            loss = ...
-            average.append(loss)
-            if i % 5 == 0:
-                print(f"cumulative average of loss: {average.aggregate()}")
-        average.reset()
+        from monai.metrics import CumulativeAverage
+
+        run_avg = CumulativeAverage()
+        for i in range(len(train_set)):
+            ...
+            val = calc_metric(x,y) #some metric value
+            run_avg.append(val)
+
+        val_avg = run_avg.aggregate() #average value
 
     """
 
-    def __init__(self) -> None:
-        super().__init__()
-        self.sum = None
-        self.not_nans = None
+    def __init__(self, group: Optional[ProcessGroup] = None) -> None:
+        """
+        Args:
+            group: Optional process group to use in multi-gpu distributed setup. Defaults to None
+            to use the default process group created during DistributedDataParallel initialization.
+        """
+
+        self.is_distributed = dist.is_available() and dist.is_initialized()
+        self.device = None
+
+        if self.is_distributed:
+            self.group = group
+            if dist.get_backend(group=group) == dist.Backend.NCCL:
+                self.device = torch.device("cuda", dist.get_rank(group=group))
+        else:
+            self.group = None
+
+        self.reset()
+
+    def reset(self) -> None:
+        """
+        Reset all the running status
+
+        """
+        self.val: torch.Tensor = 0  # type: ignore
+        self.sum: torch.Tensor = 0  # type: ignore
+        self.count: torch.Tensor = 0  # type: ignore
 
-    def reset(self):
+    def proper_tensor(self, a: Any) -> torch.Tensor:
         """
-        Reset all the running status, including buffers, sum, not nans count, etc.
+        Ensure torch.Tensor float format and optionally copy to the proper gpu device
+        """
+        x: torch.Tensor = a.detach() if isinstance(a, torch.Tensor) else torch.tensor(a)
+
+        if self.device is not None:
+            x = x.to(device=self.device, dtype=torch.float)
+        else:
+            x = x.cpu().float()
```

Comment:
If `self.device` isn't ever `None`:
```suggestion
        x = x.to(device=self.device, dtype=torch.float)
```

Reply:
updated